src/esp/gfx/RenderingTarget.cpp

```diff
@@ -0,0 +1,190 @@
+                            framebufferSize_);
+    depthBuffer_.setStorage(GL::RenderbufferFormat::R32F, framebufferSize_);
+    objectIdBuffer_.setStorage(GL::RenderbufferFormat::R32UI, framebufferSize_);
+    depthRenderbuffer_.setStorage(GL::RenderbufferFormat::Depth24Stencil8,
+                                  framebufferSize_);
+
+    framebuffer_ = GL::Framebuffer{{{}, framebufferSize_}};
+    framebuffer_
+        .attachRenderbuffer(GL::Framebuffer::ColorAttachment{0}, colorBuffer_)
+        .attachRenderbuffer(GL::Framebuffer::ColorAttachment{1}, depthBuffer_)
+        .attachRenderbuffer(GL::Framebuffer::ColorAttachment{2},
+                            objectIdBuffer_)
+        .attachRenderbuffer(GL::Framebuffer::BufferAttachment::Depth,
+                            depthRenderbuffer_)
+        .mapForDraw({{0, GL::Framebuffer::ColorAttachment{0}},
+                     {1, GL::Framebuffer::ColorAttachment{1}},
+                     {2, GL::Framebuffer::ColorAttachment{2}}});
+    CORRADE_INTERNAL_ASSERT(
+        framebuffer_.checkStatus(GL::FramebufferTarget::Draw) ==
+        GL::Framebuffer::Status::Complete);
+  }
+
+  void renderEnter() {
+    framebuffer_.bind();
+    framebuffer_.clear(GL::FramebufferClear::Color |
+                       GL::FramebufferClear::Depth);
+    framebuffer_.clearColor(1, Vector4{});
+    framebuffer_.clearColor(2, Vector4ui{});
+    framebuffer_.bind();
+  }
+
+  void renderExit() { glBindFramebuffer(GL_FRAMEBUFFER, 0); }
+
+  gltensor::GLTensorParam::ptr glTensorParam() const {
+    auto param = std::make_shared<gltensor::GLTensorParam>();
+
+    param->height_ = framebufferSize_[1];
+    param->width_ = framebufferSize_[0];
+
+    param->target_ = GL_RENDERBUFFER;
+    param->device_id_ = context_->gpuDevice();
+
+    return param;
+  }
+
+  gltensor::GLTensorParam::ptr glTensorParamRgba() const {
+    auto param = this->glTensorParam();
+
+    param->format_ = GL_RGBA;
+    param->image_ = colorBuffer_.id();
+    param->channels_ = 4;
+
+    return param;
+  }
+
+  gltensor::GLTensorParam::ptr glTensorParamDepth() const {
+    auto param = this->glTensorParam();
+
+    param->format_ = GL_R32F;
+    param->image_ = depthBuffer_.id();
+    param->channels_ = 1;
+
+    return param;
+  }
+
+  gltensor::GLTensorParam::ptr glTensorParamId() const {
+    auto param = this->glTensorParam();
+
+    param->format_ = GL_R32UI;
+    param->image_ = objectIdBuffer_.id();
+    param->channels_ = 1;
+
+    return param;
+  }
+
+  void readFrameRgba(uint8_t* ptr) {
+    framebuffer_.mapForRead(GL::Framebuffer::ColorAttachment{0});
+    Image2D rgbaImage =
+        framebuffer_.read(Range2Di::fromSize({0, 0}, framebufferSize_),
+                          {PixelFormat::RGBA8Unorm});
+    uint8_t* src_ptr = rgbaImage.data<uint8_t>();
+    std::memcpy(
+        ptr, src_ptr,
+        framebufferSize_[0] * framebufferSize_[1] * 4 * sizeof(uint8_t));
```

Comment:
Um, I must have missed this. Gimme a while, I'll give you an API that allows you to read directly into `ptr` without doing an extra allocation and copy (and without needing to go through Eigen on the bindings side). 

(It could be done with `Image2D` aready with a bunch of hacks, assumptions, wishes, and custom deleters, but it's far from pretty.)

Reply:
Maybe just Image2D -> python buffer?  It'll need to get into a numpy array at some point :)

Reply:
What I have currently in-progress is reading directly into a preallocated numpy array (so fully zero-copy). `Image2D -> buffer` would still be at least one copy.

I'll submit this in a separate PR once I'm done.